crates/sui-indexer/src/handlers/checkpoint_handler.rs

```diff
@@ -840,129 +868,116 @@ where
-                    IndexerError::FullNodeReadingError(format!(
-                        "Failed to get latest system state with error {:?}",
-                        e
-                    ))
-                })?;
-            let validators = system_state
-                .active_validators
-                .iter()
-                .map(|v| (system_state.epoch, v.clone()).into())
-                .collect();
-
-            Some(TemporaryEpochStore {
-                last_epoch: None,
-                new_epoch: DBEpochInfo {
-                    epoch: 0,
-                    first_checkpoint_id: 0,
-                    epoch_start_timestamp: system_state.epoch_start_timestamp_ms as i64,
-                    ..Default::default()
-                },
-                system_state: system_state.into(),
-                validators,
-            })
-        } else if let Some(end_of_epoch_data) = &checkpoint.end_of_epoch_data {
-            // Find system state object
-            let system_state: SuiSystemStateSummary = self
-                .http_client
-                .get_latest_sui_system_state()
-                .await
-                .map_err(|e| {
-                    IndexerError::FullNodeReadingError(format!(
-                        "Failed to get latest system state with error {:?}",
-                        e
-                    ))
-                })?;
-
-            let epoch_event = transactions.iter().find_map(|tx| {
-                tx.events.data.iter().find(|ev| {
-                    ev.type_.address == SUI_SYSTEM_ADDRESS
-                        && ev.type_.module.as_ident_str() == ident_str!("sui_system_state_inner")
-                        && ev.type_.name.as_ident_str() == ident_str!("SystemEpochInfoEvent")
-                })
-            });
-
-            let event = epoch_event
-                .map(|e| bcs::from_bytes::<SystemEpochInfoEvent>(&e.bcs))
-                .transpose()?;
-
-            let validators = system_state
-                .active_validators
-                .iter()
-                .map(|v| (system_state.epoch, v.clone()).into())
-                .collect();
-
-            let epoch_commitments = end_of_epoch_data
-                .epoch_commitments
-                .iter()
-                .map(|c| match c {
-                    CheckpointCommitment::ECMHLiveObjectSetDigest(d) => {
-                        Some(d.digest.into_inner().to_vec())
-                    }
+        // NOTE: Index epoch when object index has reached the same checkpoint,
+        // because epoch info is based on the latest system state by the current checkpoint.
+        let mut epoch_index = None;
+        if index_epoch {
+            epoch_index = if checkpoint.epoch == 0 && checkpoint.sequence_number == 0 {
+                // very first epoch
+                let system_state = get_sui_system_state(data)?;
+                let system_state: SuiSystemStateSummary = system_state.into_sui_system_state_summary();
```

Comment:
@kristaki90 they need to understand codes around here and backfill indexer from genesis.